tctl: add edit command

zmb3 opened the pull request:
This command allows you to modify a resource in place by opening the resource YAML in your text editor.

The editor is selected by checking the following, in order of precedence:

- the TELEPORT_EDITOR environment variable
- the VISUAL environment variable
- the EDITOR environment variable
- defaulting to 'vi'

zmb3 committed:
tctl: add edit command

This command allows you to modify a resource in place by opening
the resource YAML in your text editor.

The editor is selected by checking the following, in order of
precedence:

- the TELEPORT_EDITOR environment variable
- the VISUAL environment variable
- the EDITOR environment variable
- defaulting to 'vi'

espadolini commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,118 @@
+		return true, trace.BadParameter("skipping resource update, editor did not complete successfully: %v", err)
+	}
+
+	// TODO(zmb3): consider skipping this step if the file hasn't changed
```
I think unmodified and empty/only whitespace/only comments should result in an abort of the edit.

zmb3 replied:
How strongly do you feel about this? It's a decent amount of extra code to detect modifications since we just have a generic YAML file here, and aborting the edit is easily accomplished by closing the file without saving.

espadolini commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,118 @@
+
+// editor gets the text editor to be used for editing the resource
+func editor() string {
+	for _, v := range []string{"TELEPORT_EDITOR", "VISUAL", "EDITOR"} {
```
`VISUAL` before `EDITOR`? Is that what's usually done?

zmb3 replied:
Yes.

espadolini commented on tool/tctl/common/resource_command.go:
```diff
@@ -642,6 +644,76 @@ func (rc *ResourceCommand) createNode(ctx context.Context, client auth.ClientI,
 	return trace.Wrap(err)
 }
 
+func (rc *ResourceCommand) createConnector(ctx context.Context, client auth.ClientI, raw services.UnknownResource) error {
```
Why is this function responsible for two different resource kinds?

zmb3 replied:
A question for the history books, I suppose. I wouldn't want to rewrite the SAML connector logic as part of a PR that introduces a new command, so if we want to revisit this later we can. It hasn't been a problem to date so I don't think it's urgent.

espadolini commented on tool/tctl/common/resource_command.go:
```diff
@@ -642,6 +644,76 @@ func (rc *ResourceCommand) createNode(ctx context.Context, client auth.ClientI,
+		if err != nil && !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		exists = (err == nil)
```
Should we add proper CreateSAMLConnector and CreateOIDCConnector APIs?

espadolini commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,118 @@
+
+	args := strings.Fields(editor())
+	editorCmd := exec.CommandContext(ctx, args[0], append(args[1:], f.Name())...)
+	editorCmd.Env = os.Environ()
```
A `nil` `Env` will do this automatically.

espadolini commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,118 @@
+	editorCmd := exec.CommandContext(ctx, args[0], append(args[1:], f.Name())...)
+	editorCmd.Env = os.Environ()
+	editorCmd.Stdin = os.Stdin
+	editorCmd.Stdout = os.Stdout
```
```suggestion
	editorCmd.Stdout = os.Stdout
	editorCmd.Stderr = os.Stderr
```

camscale commented on tool/tctl/common/resource_command.go:
```diff
@@ -276,7 +278,7 @@ func (rc *ResourceCommand) Create(ctx context.Context, client auth.ClientI) (err
 		if !found {
 			// if we're trying to create an OIDC/SAML connector with the OSS version of tctl, return a specific error
 			if raw.Kind == "oidc" || raw.Kind == "saml" {
```
This branch is never taken now that `rc.CreateHandlers` has entries for SAML and OIDC. We emit a less useful error message now, and given a recent change to the SAML error message but not yet to OIDC, the SAML errors are also yuck:

```
$ tctl create -f r/connector-okta-dev.yaml
ERROR: rpc error: code = Unknown desc = SAML: this feature requires Teleport Enterprise

$ tctl create -f r/connector-oidc-auth0.yaml
ERROR: OIDC is only available in enterprise subscriptions
```

I'm trying to figure out where all the wrapped errors get converted to a GRPC status and why we get different error messages for each of these. Underneath they both wrap `trace,AccessDenied`, but OIDC returns the `trace.AccessDenied` error directly whereas SAML uses `trace.Wrap` on it.

If we want to still output this specific error message, we're going to have to either put it in the server or identify AccessDenied errors when trying to create these resources.

zmb3 replied:
Finally got back to this today.

Unfortunately, the `trace` package has a number of nuances here that prevent us from being able to propagate wrapped errors over the gRPC interface.

I opted to preserve the clean user-facing error message (without the confusing RPC error stuff in there), even though it means we can't wrap `ErrRequiresEnterprise` for now.

zmb3 committed:
Remove note about enterprise only

The goal is to get rid of the enterprise tctl binary and enforce
enterprise checks server-side only.

mdwn reviewed (approved):
Pending the rest of the comments, this LGTM!

zmb3 committed:
Clean up error messages

Unfortunately, the trace package and its gRPC support don't
allow us to correctly propagate wrapped errors across the
gRPC connection. For now, prefer a clear and concise user-facing
message. We can investigate improving the trace package's wrapping
functionality in the future (it was developed before many of Go's
error wrapping features existed).

zmb3 committed:
Fix imports

zmb3 committed:
Prevent renaming resources with tctl edit

See also gravitational/webapps#1465 where we do the same
for the web UI.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -2769,7 +2769,10 @@ func (a *ServerWithRoles) UpsertOIDCConnector(ctx context.Context, connector typ
 	}
 	if !modules.GetModules().Features().OIDC {
-		return trace.AccessDenied("OIDC is only available in enterprise subscriptions")
+		// TODO(zmb3): ideally we would wrap ErrREquiresEnterprise here, but
+		// we can't currently propagate wrapped errors across the gRPC boundary,
+		// and we want tctl to display a clean user-facing messasge in this case
```
```suggestion
		// TODO(zmb3): ideally we would wrap ErrRequiresEnterprise here, but
		// we can't currently propagate wrapped errors across the gRPC boundary,
		// and we want tctl to display a clean user-facing message in this case
```

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -2846,15 +2849,20 @@ func (a *ServerWithRoles) DeleteOIDCConnector(ctx context.Context, connectorID s
 // UpsertSAMLConnector creates or updates a SAML connector.
 func (a *ServerWithRoles) UpsertSAMLConnector(ctx context.Context, connector types.SAMLConnector) error {
+	if !modules.GetModules().Features().SAML {
+		// TODO(zmb3): ideally we would wrap ErrREquiresEnterprise here, but
+		// we can't currently propagate wrapped errors across the gRPC boundary,
+		// and we want tctl to display a clean user-facing messasge in this case
```
```suggestion
		// TODO(zmb3): ideally we would wrap ErrRequiresEnterprise here, but
		// we can't currently propagate wrapped errors across the gRPC boundary,
		// and we want tctl to display a clean user-facing message in this case
```

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -2846,15 +2849,20 @@ func (a *ServerWithRoles) DeleteOIDCConnector(ctx context.Context, connectorID s
+		// TODO(zmb3): ideally we would wrap ErrREquiresEnterprise here, but
+		// we can't currently propagate wrapped errors across the gRPC boundary,
+		// and we want tctl to display a clean user-facing messasge in this case
+		return trace.Wrap(ErrSAMLRequiresEnterprise)
```
The comment above says that "ideally we'd wrap ErrRequiresEnterprise" but this is exactly what we're doing here, no?

r0mant commented on lib/auth/saml.go:
```diff
@@ -31,7 +30,11 @@ import (
 // using the Enterprise edition of Teleport.
-var ErrSAMLRequiresEnterprise = fmt.Errorf("SAML: %w", ErrRequiresEnterprise)
+//
+// TODO(zmb3): ideally we would wrap ErrRequiresEnterprise here, but
+// we can't currently propagate wrapped errors across the gRPC boundary,
+// and we want tctl to display a clean user-facing messasge in this case
```
Nit: This comment feels out of place on this variable declaration tbh, or maybe I misunderstood.

zmb3 replied:
I don't want someone to accidentally change this, because `trace` is very finicky with gRPC. If you wrap the error in any way, the user-facing error message that tctl prints will have extra RPC garbage in it.

r0mant commented on tool/tctl/common/resource_command.go:
```diff
@@ -110,6 +110,8 @@ func (rc *ResourceCommand) Initialize(app *kingpin.Application, config *service.
 		types.KindToken:                   rc.createToken,
 		types.KindInstaller:               rc.createInstaller,
 		types.KindNode:                    rc.createNode,
+		types.KindOIDCConnector:           rc.createOIDCConnector,
+		types.KindSAMLConnector:           rc.createSAMLConnector,
```
Are these changes part of the effort to get rid of enterprise-specific tctl?

zmb3 replied:
Kind of. They're necessary because we're leveraging `tctl`'s `ResourceCommand` to get the existing resource.

If we keep today's approach of having a `ResourceCommand` and a ResourceCommandE`, then we would have to create an `EditCommand` and an `EditCommandE`.

Since we're trying to get rid of enterprise tctl, I consolidated on a single resource command. This means you can now create OIDC and SAML connectors with OSS tctl or enterprise tctl, so long as the auth server is enterprise.

r0mant commented on tool/tctl/common/resource_command.go:
```diff
@@ -644,6 +642,66 @@ func (rc *ResourceCommand) createNode(ctx context.Context, client auth.ClientI,
 	return trace.Wrap(err)
 }
 
+func (rc *ResourceCommand) createOIDCConnector(ctx context.Context, client auth.ClientI, raw services.UnknownResource) error {
```
Fetching and deleting OIDC/SAML connectors is already in open-source tctl, right?

zmb3 replied:
Yeah, it kind of always "just worked" because we only explicitly check the kind on create calls.

r0mant commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	// nothing to do if the resource was not modified
+	if changed := newSum != originalSum; !changed {
```
Why not just `if newSum == originalSum`?

r0mant commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,178 @@
+			return value
+		}
+	}
+	return "vi"
```
Should be `emacs` 😁

r0mant commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,178 @@
+
+// editor gets the text editor to be used for editing the resource
+func editor() string {
+	for _, v := range []string{"TELEPORT_EDITOR", "VISUAL", "EDITOR"} {
```
Nit: Move `TELEPORT_EDITOR` into a constant e.g. like we do [in tsh](https://github.com/gravitational/teleport/blob/master/tool/tsh/tsh.go#L479-L484).

zmb3 replied:
I kind of like that all 3 literals are right here, so you don't have to go searching around for them and you can see the order of precedence. Since `TELEPORT_EDITOR` isn't used anywhere else I don't see much value in the constant.

r0mant commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,178 @@
+
+	// nothing to do if the resource was not modified
+	if changed := newSum != originalSum; !changed {
+		return true, nil
```
Should we give some feedback to the user if the resource didn't change? I think `kubectl edit` prints something in the console in this case.

zmb3 replied:
yep, good call. just confirmed kubectl does print a message here.

r0mant commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,178 @@
+	if err := rc.Create(ctx, client); err != nil {
+		return true, trace.Wrap(err)
+	}
+
```
Same as above, I'd print something to the console to indicate the resource was successfully edited.

zmb3 replied:
This is done inside `rc.Create`, so no action necessary.

r0mant commented on tool/tctl/common/edit_command.go:
```diff
@@ -0,0 +1,178 @@
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
```
Nit: Wonder if we should trim whitespaces before calculating the checksum.

zmb3 replied:
I'm not too worried about it. If nothing but whitespace changed and we make an extra call to the backend that ends up creating the same resource, oh well.

r0mant reviewed (approved):
lgtm with a few suggestions

zmb3 committed:
Address review comments

zmb3 committed:
Fix misspelling
